src/include/cursor.i

```diff
@@ -17,6 +17,55 @@ __cursor_set_recno(WT_CURSOR_BTREE *cbt, uint64_t v)
     cbt->iface.recno = cbt->recno = v;
 }
 
+/*
+ * __cursor_free_item --
+ *     Release memory used by the key or value item in cursor copy debug mode.
+ */
+static inline int
+__cursor_free_item(WT_SESSION_IMPL *session, WT_ITEM *item)
+{
+    void *copy_addr;
+
+    if (item->mem != NULL && F_ISSET(S2C(session), WT_CONN_DEBUG_CURSOR_COPY)) {
+        /*
+         * If we own the memory for the item, make a copy and point to the copy. That allows us to
+         * overwrite and free the original memory, potentially uncovering programming errors related
+         * to retaining pointers to key/value memory beyond API boundaries. If the item has memory
+         * not currently in use, free that as well.
+         */
+        if (WT_DATA_IN_ITEM(item)) {
+            WT_RET(__wt_memdup(session, item->mem, item->memsize, &copy_addr));
+            __wt_explicit_overwrite(item->mem, item->memsize);
+            __wt_free(session, item->mem);
+            item->data = item->mem = copy_addr;
+        } else {
+            __wt_explicit_overwrite(item->mem, item->memsize);
+            __wt_buf_free(session, item);
+        }
+    }
+    return (0);
```

Comment:
It's not pretty and it's slower, but I think we'll be safer if we use the buffer functions. For example, you're not handling the possibility that `item.data != item.mem` (which can happen when reading overflow records, for example), and nobody is going to want to debug that at some future date.

How much do you hate something like this?
```
{
    WT_DECL_item(tmp);
    WT_DECL_RET;
   
    if (!F_ISSET(S2C(session), WT_CONN_DEBUG_CURSOR_COPY))  
        return;

    if (WT_DATA_IN_ITEM(item)) {
        WT_ERR(__wt_scr_alloc(session, 0, &tmp));
        WT_ERR(__wt_buf_set(session, tmp, item->data, item->size));
        __wt_overwrite_and_free_len(session, item->mem, item->memsize);
        __wt_buf_free(session, item);
        WT_ERR(__wt_buf_set(session, item, tmp->data, tmp->size));
    } else {
        __wt_overwrite_and_free_len(session, item->mem, item->memsize);
        __wt_buf_free(session, item);
    }
err:
    __wt_scr_free(session, &tmp);
    return (ret);
} 
```

Reply:
I did go this route.  But I noticed that as soon as the memory used by the key/data is freed, it is used again.  I ended up adding yet another allocation so this doesn't happen (at least on this version of malloc or others that do some sort of LIFO allocation for similar sizes).  I added a comment to explain.